contracts/Staking.sol

```diff
@@ -74,18 +77,22 @@ contract Staking is Ownable {
     @param tokenAddress Token address
     @param data should be disregarded for the current deployment
      */
-    function stakeFor(uint256 amount, address user, address tokenAddress, bytes memory data) public payable {
-        require(isInWhiteList(tokenAddress), "Staking: not possible to interact with this token");
+    function stakeFor(
+        uint256 amount,
+        address user,
+        address tokenAddress,
+        bytes memory data
+    ) public payable whitelistedToken(tokenAddress) {
         // disregard passed-in amount
-        if(_isNativeToken(tokenAddress)) {
+        if (_isNativeToken(tokenAddress)) {
             amount = msg.value;
-            tokenAddress = address(0);
-        } else {
-            IERC20(tokenAddress).safeTransferFrom(msg.sender, address(this), amount);
         }
         _amountStaked[user][tokenAddress] = _amountStaked[user][tokenAddress].add(amount);
         _totalStaked[tokenAddress] = _totalStaked[tokenAddress].add(amount);
         emit Staked(user, amount, _amountStaked[user][tokenAddress], tokenAddress, data);
+        if (!_isNativeToken(tokenAddress)) {
+            IERC20(tokenAddress).safeTransferFrom(msg.sender, address(this), amount);
+        }
```

Comment:
I think the message from the security guys is that we make sure no state changes are made after an external call. The event triggering however should go last. It's a notification rather than a state change, and if the external call fails - we will mislead the listening party.

https://consensys.github.io/smart-contract-best-practices/recommendations/#avoid-state-changes-after-external-calls
https://docs.soliditylang.org/en/develop/security-considerations.html?highlight=check%20effects#use-the-checks-effects-interactions-pattern

In the `Checks-Effects-Interactions` pattern event firing is rather interaction than effect.
Let me know what you think about it. Thanks.

Reply:
the same comment is fare to all other functions as well where an event is triggered

Reply:
Slither show alerts if we emit events after the external call, that's only the reason why oi change it

Reply:
I think its OK to emit at the end, there shouldnt be issues if we emit after call as Reentrancy occurs during same transaction and events cannot be consumed by smart contracts so it has no impact at all.

Reply:
I think its clearear and common practice to emit at the end.

Reply:
I wonder why slither warns us then? Maybe it cannot detect the nature of the code and see the different between the state changes and not state changes..

Reply:
Done